apps/dash-study-browser/app.py

```diff
@@ -0,0 +1,322 @@
+
+    if study is None:
+        study = study_data.study_id[0]
+
+    study_data = study_data[study_data.study_id == study]
+    vehicle_readings = study_data["reading_value"][
+        study_data["group_type"] == "control"
+    ]
+    data_range = study_data["reading_value"].max() - study_data["reading_value"].min()
+
+    test_stats = {}
+    box_data = []
+    violin_data = []
+    for i, group_id in enumerate(study_data.group_id.unique()):
+        try:
+            group_name = study_data["group_name"][
+                study_data.group_id == group_id
+            ].values[0]
+        except KeyError:
+            group_name = group_id
+
+        group_type = study_data["group_type"][study_data.group_id == group_id].values[0]
+        y_data = study_data["reading_value"][study_data.group_id == group_id]
+
+        try:
+            subject_ids = study_data["subject_id"][study_data.group_id == group_id]
+        except KeyError:
+            subject_ids = None
+
+        t, p = stats.ttest_ind(vehicle_readings, y_data)
+        test_stats[group_id] = {
+            "t": t,
+            "p": p,
+            "pf": "p={:0.3f}".format(p) if p >= 0.001 else "p<0.001",
+            "astrix": "***"
+            if p <= 0.001
+            else "**"
+            if p <= 0.01
+            else "*"
+            if p <= 0.05
+            else "",
+            "max_y": y_data.max(),
+            "index": i,
+        }
+
+        box_data.append(
+            go.Box(
+                y=y_data,
+                name=group_name,
+                text=subject_ids,
+                hoveron="points",
+                boxmean=True,
+                showlegend=False,
+                boxpoints="all",
+                pointpos=0,
+                line={"color": group_colors.get(group_type, "green")},
+            )
+        )
+        violin_data.append(
+            go.Violin(
+                y=y_data,
+                name=group_name,
+                text=subject_ids,
+                hoveron="points",
+                meanline={"visible": True},
+                showlegend=False,
+                points="all",
+                pointpos=0,
+                line={"color": group_colors.get(group_type, "green")},
+            )
+        )
+
+    chart_data = {"box": box_data, "violin": violin_data}
+
+    if "reading_name" in study_data.columns:
+        reading_name = study_data["reading_name"].unique()[0]
+    else:
+        reading_name = None
+
+    if not vehicle_readings.empty:
+        ref_groups = set(
+            study_data.group_id[study_data.group_type == "reference"].unique()
+        )
+        control_groups = set(
+            study_data.group_id[study_data.group_type == "control"].unique()
+        )
+        all_groups = set(study_data.group_id.unique())
+        groups_to_annotate = all_groups - ref_groups - control_groups
+        annotations = [
+            dict(
+                x=test_stats[group_id]["index"],
```

Comment:
if `study_data.group_id` is empty, it won't loop and add data to the `tests_stats`.

This will cause a index error `test_stats[group_id]["index"]`. I'm not sure if this edge case will appear often.

Reply:
I don't think this case will happen, because `group_id` is a necessary prereq for the study data

Reply:
It's a good idea to program defensively! You can use the `.get()` method for Python dictionaries  to return a default value if `group_id` isn't in `test_stats`.

Reply:
I have a test in the error message callback to see if `group_id  in the headers is in the csv file. If not then use a default data set.
```
            missing_columns = {
                "group_id",
                "group_type",
                "reading_value",
                "study_id",
            }.difference(study_data.columns)
```

Also if the csv file has the headers, but is empty the `plot` figure callback should return an empty dictionary for the figure. 

But I added the `.get()` check anyway just to be sure.